File: backend/app/api/compare.py
import shutil
import uuid
from pathlib import Path

from fastapi import APIRouter, HTTPException
from fastapi.responses import RedirectResponse

from app.config import TEMP_DIR
from app.models.schemas import CompareResponse, CompareURLRequest
from app.services import history_db
from app.services.diff_engine import diff_lines
from app.services.line import blocks_to_lines
from app.services.mapper import build_compare_result
from app.services.minio_client import download_bytes
from app.services.pdf_extract import extract_text_blocks

router = APIRouter(prefix="/api", tags=["compare"])


@router.post("/compare", response_model=CompareResponse)
async def compare_pdfs(body: CompareURLRequest):
    try:
        template_content, template_type = download_bytes(body.template_url)
        contract_content, contract_type = download_bytes(body.contract_url)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    if template_type not in {"application/pdf", "application/octet-stream"}:
        raise HTTPException(status_code=400, detail="模版文件必须是 PDF")
    if contract_type not in {"application/pdf", "application/octet-stream"}:
        raise HTTPException(status_code=400, detail="正式文件必须是 PDF")

    compare_options = body.options
    job_id = str(uuid.uuid4())
    job_dir = TEMP_DIR / job_id
    job_dir.mkdir(parents=True, exist_ok=True)

    template_path = job_dir / "template.pdf"
    contract_path = job_dir / "contract.pdf"

    try:
        template_path.write_bytes(template_content)
        contract_path.write_bytes(contract_content)

        template_blocks = extract_text_blocks(
            template_path,
            ignore_header_footer=compare_options.ignore_header_footer,
        )
        contract_blocks = extract_text_blocks(
            contract_path,
            ignore_header_footer=compare_options.ignore_header_footer,
        )

        template_lines = blocks_to_lines(
            template_blocks,
            prefix="tpl",
            ignore_whitespace=compare_options.ignore_whitespace,
            filter_irrelevant=compare_options.ignore_header_footer,
        )
        contract_lines = blocks_to_lines(
            contract_blocks,
            prefix="con",
            ignore_whitespace=compare_options.ignore_whitespace,
            filter_irrelevant=compare_options.ignore_header_footer,
        )

        raw_changes = diff_lines(template_lines, contract_lines)

        result = build_compare_result(
            job_id,
            template_lines,
            contract_lines,
            raw_changes,
        )

        try:
            history_db.save_history(
                job_id=job_id,
                template_url=body.template_url,
                contract_url=body.contract_url,
                template_name=body.template_name,
                contract_name=body.contract_name,
                result=result,
            )
        except Exception:
            pass

        return CompareResponse(job_id=job_id, status="done", result=result)
    except HTTPException:
        raise
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"比对失败: {exc}") from exc
    finally:
        shutil.rmtree(job_dir, ignore_errors=True)


@router.get("/compare/{job_id}", response_model=CompareResponse)
async def get_compare_result(job_id: str):
    try:
        detail = history_db.get_history_by_job_id(job_id)
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"读取比对结果失败: {exc}") from exc

    if detail is None or detail.result is None:
        raise HTTPException(status_code=404, detail="任务不存在或已过期")

    return CompareResponse(job_id=job_id, status="done", result=detail.result)


@router.get("/files/{job_id}/{which}")
async def get_pdf_file(job_id: str, which: str):
    if which not in {"template", "contract"}:
        raise HTTPException(status_code=400, detail="which 只能是 template 或 contract")

    try:
        detail = history_db.get_history_by_job_id(job_id)
    except Exception as exc:
        raise HTTPException(status_code=500, detail=f"读取文件信息失败: {exc}") from exc

    if detail is None:
        raise HTTPException(status_code=404, detail="文件不存在")

    pdf_url = detail.template_url if which == "template" else detail.contract_url
    return RedirectResponse(url=pdf_url, status_code=307)

File: backend-go/internal/service/compare.go
package service

import (
	"encoding/json"
	"io"
	"os"
	"path/filepath"

	"github.com/google/uuid"
	"github.com/pdfdiff/backend-go/internal/apperror"
	"github.com/pdfdiff/backend-go/internal/config"
	"github.com/pdfdiff/backend-go/internal/dto"
	"github.com/pdfdiff/backend-go/internal/storage"
)

var allowedContentTypes = map[string]struct{}{
	"application/pdf":      {},
	"application/octet-stream": {},
}

type CompareService struct {
	cfg     config.Config
	history *storage.HistoryStore
	minio   *storage.MinioClient
}

func NewCompareService(cfg config.Config, history *storage.HistoryStore, minio *storage.MinioClient) *CompareService {
	return &CompareService{cfg: cfg, history: history, minio: minio}
}

type UploadedFile struct {
	Content     []byte
	ContentType string
}

func (s *CompareService) CompareFromURLs(req dto.CompareURLRequest) (dto.CompareResponse, error) {
	templateContent, templateType, err := s.minio.Download(req.TemplateURL)
	if err != nil {
		return dto.CompareResponse{}, err
	}
	contractContent, contractType, err := s.minio.Download(req.ContractURL)
	if err != nil {
		return dto.CompareResponse{}, err
	}

	template := UploadedFile{Content: templateContent, ContentType: templateType}
	contract := UploadedFile{Content: contractContent, ContentType: contractType}

	options := req.Options
	if options == (dto.CompareOptions{}) {
		options = dto.CompareOptions{IgnoreWhitespace: true, IgnoreHeaderFooter: true}
	}
	optionsJSON, err := json.Marshal(options)
	if err != nil {
		return dto.CompareResponse{}, apperror.BadRequest("options 参数无效")
	}

	resp, result, err := s.compareFiles(template, contract, string(optionsJSON))
	if err != nil {
		return dto.CompareResponse{}, err
	}

	if s.history != nil {
		_ = s.history.SaveHistory(
			resp.JobID,
			req.TemplateURL,
			req.ContractURL,
			req.TemplateName,
			req.ContractName,
			result,
		)
	}
	return resp, nil
}

func (s *CompareService) Compare(template, contract UploadedFile, optionsJSON string) (dto.CompareResponse, error) {
	resp, _, err := s.compareFiles(template, contract, optionsJSON)
	return resp, err
}

func (s *CompareService) compareFiles(template, contract UploadedFile, optionsJSON string) (dto.CompareResponse, dto.CompareResult, error) {
	if err := validatePDF(template.ContentType, "模版文件必须是 PDF"); err != nil {
		return dto.CompareResponse{}, dto.CompareResult{}, err
	}
	if err := validatePDF(contract.ContentType, "正式文件必须是 PDF"); err != nil {
		return dto.CompareResponse{}, dto.CompareResult{}, err
	}

	options, err := parseOptions(optionsJSON)
	if err != nil {
		return dto.CompareResponse{}, dto.CompareResult{}, err
	}

	jobID := uuid.New().String()
	jobDir := filepath.Join(s.cfg.TempDir, jobID)

	if err := os.MkdirAll(jobDir, 0o755); err != nil {
		return dto.CompareResponse{}, dto.CompareResult{}, apperror.Internal("比对失败: " + err.Error())
	}
	defer func() { _ = os.RemoveAll(jobDir) }()

	templatePath := filepath.Join(jobDir, "template.pdf")
	contractPath := filepath.Join(jobDir, "contract.pdf")

	if err := saveUpload(template.Content, templatePath, s.cfg.MaxUploadSize); err != nil {
		return dto.CompareResponse{}, dto.CompareResult{}, err
	}
	if err := saveUpload(contract.Content, contractPath, s.cfg.MaxUploadSize); err != nil {
		return dto.CompareResponse{}, dto.CompareResult{}, err
	}

	templateBlocks, err := ExtractTextBlocks(templatePath, options.IgnoreHeaderFooter)
	if err != nil {
		return dto.CompareResponse{}, dto.CompareResult{}, apperror.Internal("比对失败: " + err.Error())
	}
	contractBlocks, err := ExtractTextBlocks(contractPath, options.IgnoreHeaderFooter)
	if err != nil {
		return dto.CompareResponse{}, dto.CompareResult{}, apperror.Internal("比对失败: " + err.Error())
	}

	templateLines := BlocksToLines(templateBlocks, "tpl", options.IgnoreWhitespace)
	contractLines := BlocksToLines(contractBlocks, "con", options.IgnoreWhitespace)
	rawChanges := DiffLines(templateLines, contractLines)
	result := BuildCompareResult(jobID, templateLines, contractLines, rawChanges)

	return dto.DoneResponse(jobID, result), result, nil
}

func (s *CompareService) MaxUploadSize() int64 {
	return s.cfg.MaxUploadSize
}

func (s *CompareService) GetResult(jobID string) (dto.CompareResponse, error) {
	if s.history == nil {
		return dto.CompareResponse{}, apperror.NotFound("任务不存在或已过期")
	}

	detail, err := s.history.GetHistoryByJobID(jobID)
	if err != nil {
		return dto.CompareResponse{}, apperror.Internal("读取比对结果失败: " + err.Error())
	}
	if detail == nil {
		return dto.CompareResponse{}, apperror.NotFound("任务不存在或已过期")
	}
	return dto.DoneResponse(jobID, detail.Result), nil
}

func parseOptions(optionsJSON string) (dto.CompareOptions, error) {
	options := dto.CompareOptions{
		IgnoreWhitespace:   true,
		IgnoreHeaderFooter: true,
	}
	if optionsJSON == "" {
		return options, nil
	}
	if err := json.Unmarshal([]byte(optionsJSON), &options); err != nil {
		return options, apperror.BadRequest("options 参数无效: " + err.Error())
	}
	return options, nil
}

func validatePDF(contentType, message string) error {
	if _, ok := allowedContentTypes[contentType]; !ok {
		return apperror.BadRequest(message)
	}
	return nil
}

func saveUpload(content []byte, dest string, maxSize int64) error {
	if int64(len(content)) > maxSize {
		return apperror.BadRequest("文件大小超过 50MB 限制")
	}
	return os.WriteFile(dest, content, 0o644)
}

func ReadUpload(r io.Reader, maxSize int64) ([]byte, error) {
	limited := io.LimitReader(r, maxSize+1)
	data, err := io.ReadAll(limited)
	if err != nil {
		return nil, err
	}
	if int64(len(data)) > maxSize {
		return nil, apperror.BadRequest("文件大小超过 50MB 限制")
	}
	return data, nil
}
